### PR TITLE
Add `Warning` class and refactor C++ warnings to use it

### PR DESCRIPTION
**Pull request by kurtamohler**
Also adds `TORCH_WARN_WITH` and `TORCH_WARN_DEPRECATION` macros

Part of #72948

### CODE REVIEW

File: c10/util/Exception.cpp [hunk 1]
```diff
@@ -147,25 +147,44 @@ thread_local WarningHandler* ThreadWarningHandler::warning_handler_ = nullptr;
...
-void warn(
+Warning::Warning(
     const SourceLocation& source_location,
     const std::string& msg,
-    const bool verbatim) {
-  ThreadWarningHandler::get_handler()->process(source_location, msg, verbatim);
-}
+    const bool verbatim)
+    : source_location_(std::move(source_location)),
+      msg_(std::move(msg)),
```
Comment (Lezcano) [inline c10/util/Exception.cpp:155]: You cannot move from a `const&`. If you want to move from it, you should pass these things by value. This is sometimes referred to as the "copy-and-move idiom".

File: c10/util/Exception.cpp [hunk 2]
```diff
@@ -147,25 +147,44 @@ thread_local WarningHandler* ThreadWarningHandler::warning_handler_ = nullptr;
...
 
-void warn(
+Warning::Warning(
     SourceLocation source_location,
     detail::CompileTimeEmptyString msg,
-    const bool verbatim) {
-  warn(source_location, "", verbatim);
-}
+    const bool verbatim)
+    : Warning(source_location, "", verbatim) {}
```
Comment (Lezcano) [inline c10/util/Exception.cpp:162]: Here you have the opposite issue, the source location should be moved.

**Comment by kurtamohler** [comment]
Thanks for taking a look @Lezcano!

I found an issue with the warning class just now. We need to be able to support subclasses of the base `Warning` class which each get translated to different Python warning types. However, as it's currently implemented in this PR, when a warning gets added to `vector<Warning> warning_buffer_` in `PyWarningHandler::InternalHandler::process`, the warning gets copied as a base `Warning` class. So if a subclass is being used, that information is lost, and all warnings get converted to the same Python warning type.

After some googling, it looks like I might have to store `shared_ptr`s of warnings in the `warning_buffer_`, rather than the `Warning` object directly. I'll see if that solution works.

**Comment by Lezcano** [comment]
If you want to use subclasses, then yes, you need to store pointers to them. Depending on what you want to do, you may want to store them as `unique_ptr`s, `shared_ptr`s or `weak_ptr`s of course.

File: torch/csrc/Exceptions.cpp [hunk 3]
```diff
@@ -210,18 +210,32 @@ LinAlgError::LinAlgError(const char* format, ...) {
...
 
 PyWarningHandler::PyWarningHandler() noexcept(true)
     : prev_handler_(c10::Warning::get_warning_handler()), in_exception_(false) {
   c10::Warning::set_warning_handler(&internal_handler_);
 }
 
+// Get the Python warning type for a warning
+// TODO: It would be much better to have a method like
+// `virtual PyObject* Warning::python_type()` instead of this mapping function.
+PyObject* map_warning_to_python_type(c10::Warning::Warning& warning) {
```
Comment (kurtamohler) [inline torch/csrc/Exceptions.cpp:225]: I would rather have a `PyObject* Warning::python_type()` function to replace this, but I'm not sure if it's possible to implement it. The issue is that `c10/util/Exceptions.cpp` doesn't have access to the python headers, so if I try to define the function there, I can't return `PyExc_UserWarning` and others.

I also attempted defining `PyObject* Warning::python_type()` in this file, `torch/csrc/Exception.cpp`, since it does include the python headers. That worked locally for me, but it caused build errors in CI. I'm not sure, but I'm guessing that C++-only builds don't build any files that include the python headers.

File: c10/util/Exception.cpp [hunk 4]
```diff
@@ -147,25 +147,53 @@ thread_local WarningHandler* ThreadWarningHandler::warning_handler_ = nullptr;
...
+  TORCH_INTERNAL_ASSERT(
+      false, "Cannot call `Warning::warn()`, you must use a subclass of `Warning`");
+}
+
+// TODO: I don't like that subclasses currently override this method just
+// so they can do `make_shared<subclass_type>`. Is there another way to make
+// a shared_ptr of the proper type without needing to override this method for
+// each subclass?
+void UserWarning::warn() const {
+  ThreadWarningHandler::get_handler()->process(std::make_shared<UserWarning>(*this));
```
Comment (amjames) [inline c10/util/Exception.cpp:196]: Another option is to make `warn()` a factory template. It looks like the macros are used to instantiate warning objects and `.warn()` is called immediately. So the factory becomes responsible for creating the object and immediately passing to the handler. 

```c++
template<typename WarningType, typename... Args>
void warn(Args&&... args) { 
   ThreadWarningHandler::get_handler()->process(std::make_shared<WarningType>(std::forward<Args>(args)...));
}
```

Comment (kurtamohler) [reply c10/util/Exception.cpp:196]: That's a good solution, thanks!

Comment (kurtamohler) [reply c10/util/Exception.cpp:196]: Actually, I'm not sure this works the way we need it to. This template function has to be called like `warn<WarningType>(...);`, with the warning subclass explicitly specified. That's a problem in `torch/csrc/autograd/utils/warnings.cpp`--when warnings are replayed from the buffer, they could be warnings of any subclass type. So I think we have to stick with using the `Warning::warn()`/`UserWarning::warn()` methods, since they don't need the subclass to be specified explicitly

File: c10/util/Exception.cpp [hunk 5]
```diff
@@ -116,7 +116,7 @@ void torchInternalAssertFail(
 
 } // namespace detail
 
-namespace Warning {
+namespace WarningUtils {
```
Comment (kurtamohler) [inline c10/util/Exception.cpp:119]: Not sure if I like the name `WarningUtils`, but I thought that it made sense to name the base warning class `c10::Warning`

File: c10/util/Exception.cpp (hunk 4 above)
Comment (amjames) [reply c10/util/Exception.cpp]: When replaying the warnings from the buffer you do not need to create the shared ptrs, as you already stored them. You can simply `process()` them with the appropriate handler.

File: c10/util/Exception.cpp [hunk 6]
```diff
@@ -195,14 +173,64 @@ WarnAlways::~WarnAlways() {
...
+}
+
+bool Warning::verbatim() const {
+  return verbatim_;
+}
+
+void Warning::warn() const {
+  TORCH_INTERNAL_ASSERT(
+      false,
+      "Cannot call `Warning::warn()`, you must use a subclass of `Warning`");
```
Comment (amjames) [inline c10/util/Exception.cpp:217]: Should the base warning class be abstract?

File: c10/util/Exception.cpp (hunk 4 above)
Comment (Lezcano) [reply c10/util/Exception.cpp]: For what is worth, if you have an a-priori limited set of warnings and you want to store them and process them in a uniform way, you may want to use `c10::variant` together with the `visit` method. This would save you from using pointers at all.

Comment (Lezcano) [reply c10/util/Exception.cpp]: Note that since in C++14 we don't have variadic lambdas, you'd need to implement this via your own functor. Now, if you don't want to pollute the global namespace, remember that you can define the struct to be local to the function you are using it in. This is as close as you get to variadic lambdas in C++14, I guess

File: c10/util/Exception.cpp (hunk 6 above)
Comment (kurtamohler) [reply c10/util/Exception.cpp:217]: I don't think so. According to [this page](https://en.cppreference.com/w/cpp/language/abstract_class), "Abstract types cannot be used as parameter types." But I think we need to be able to use `c10::Warning` as a parameter type

File: c10/util/Exception.cpp (hunk 4 above)
Comment (kurtamohler) [reply c10/util/Exception.cpp]: @amjames, I see, you're right. However, today I tried again to implement this template `warn` function, and I've been having some trouble getting it exactly right--I've gotten various build errors and undefined symbols at runtime. Since the template has to go in a header file, every time I change it I have to rebuild for 30+ minutes, which is a pain. I'll keep looking into it, but maybe this should be done in a follow-up PR

**Comment by kurtamohler** [comment]
@amjames brought up the question of whether we should really use polymorphic warning classes. Alternatively, we could use just one class and add an enum data member to it which corresponds with the python warning type that it will be converted to. I think the main benefits are that the implementation could be a little simpler and we wouldn't need smart pointers like the ones I added in this PR.

If we end up doing it this way, then when we add `TORCH_WARN_WITH(warning_t, ...)` in the future, per the [RFC](https://github.com/pytorch/rfcs/pull/43/files#diff-2408a781d8ee8d6167e3403c0fd8c08c23c38fa9dfedc980b98d3e892b85199bR259), then the `warning_t` would not be a warning subtype, but instead it would be one of the values in the python warning enum.

The downsides with this are the inconsistencies it would introduce between warnings and errors, as well as C++ and Python APIs.

For instance, `TORCH_WARN_WITH(warning_t, ...)` and `TORCH_CHECK_WITH(error_t, ...)` would be a little bit inconsistent with each other because `error_t` will still be an actual C++ class type. That is, unless we want to change `TORCH_CHECK_WITH` to take in an enum value too, but I don't think we'd want to do that, since having polymorphic error types allows you to `catch` specific error types.

Also, when we add the Python APIs described in the RFC, `torch.warn_with(warning_type, ...)` will take in an actual Python warning type, so that would be an inconsistency between C++ and Python.

There are probably more tradeoffs to consider.

cc @bdhirsh (I would normally ping @albanD, but he's on PTO)

File: c10/util/Exception.cpp (hunk 4 above)
Comment (kurtamohler) [reply c10/util/Exception.cpp]: Okay, I got the template function factory `c10::warn()` working and removed `c10::Warning::warn()`. I also added two more `c10::warn()` overloads that take in a `c10::Warning` by reference or a smart pointer to a `c10::Warning`--the second one helped with file scoping so that I didn't need to move the `ThreadWarningHandler` definition from `c10/utils/Exceptions.cpp` into the header file

**Comment by Lezcano** [comment]
Note that the idea in https://github.com/pytorch/pytorch/pull/84101#discussion_r965225602 of using variants is probably what you want to go with, as that's the type-safe implementation of having a class with an enum.

File: c10/util/Exception.cpp (hunk 4 above)
Comment (kurtamohler) [reply c10/util/Exception.cpp]: @Lezcano, I'm not sure if I fully understand your suggestion yet, I'm not familiar with `c10::variant`, so let me know if my interpretation is right.

It seems like part of your suggestion is that instead of, for instance, `void PyWarningHandler::InternalHandler::process(std::shared_ptr<c10::Warning> warning)` (in the `torch/csrc/Exceptions.cpp` file), we could do `void PyWarningHandler::InternalHandler::process(c10::variant<c10::UserWarning, ...> warning_variant)`. Then inside that function we would call `c10::visit([&](auto&& warning) {...}, warning_variant);`, and `warning` would have the proper subclass type inside the `{...}` scope there.

It's worth mentioning that as we add more warning subclasses, we'd have to keep adding them to `c10::variant<c10::UserWarning, ...>`. But we can just make a typedef for it like `typedef c10::variant<c10::UserWarning, ...> warning_variant_t` to be reused wherever needed

Comment (Lezcano) [reply c10/util/Exception.cpp]: That's exactly on point. The only thing is that `auto&&` is not C++14, but yep. And yes, we should make a `using variant_type = ...`

Comment (kurtamohler) [reply c10/util/Exception.cpp]: Alright, I've implemented the variant type and it seems to be working. One possible issue with it is that the `process(c10::warning_variant_t)` functions are passing the variant by value. Looks like passing a variant by reference like `process(c10::warning_variant_t&)` is not allowed. However, I could create another variant type that holds references to warnings, like `typedef c10::variant<std::reference_wrapper<UserWarning>> warning_ref_variant_t;` to use in these situations. Is it worth the extra complexity? Or is there a better way to do it?

Comment (lezcano) [reply c10/util/Exception.cpp]: I would expect that you can pass it by const reference. A variant is nothing but a glorified union with extra type-safety, so it behaves just as a normal type. What error do you get?

Comment (kurtamohler) [reply c10/util/Exception.cpp]: Oh, you're right. The error I was getting was actually because I tried to use `vector<const c10::warning_variant_t&>`. I guess it's alright to just use `vector<c10::warning_variant_t>`

Comment (lezcano) [reply c10/util/Exception.cpp]: yeah, no, a vector cannot store references.

**Comment by kurtamohler** [comment]
I think this is ready for further review

File: c10/util/Exception.h [hunk 7]
```diff
@@ -112,17 +113,56 @@ class C10_API Error : public std::exception {
...
+
+  // See note: [Verbatim Warnings]
+  bool verbatim_;
+};
+
+class C10_API UserWarning : public Warning {
+  using Warning::Warning;
+};
+
+typedef c10::variant<UserWarning> warning_variant_t;
```
Comment (albanD) [inline c10/util/Exception.h:152]: I'm not a specialist here, what is the benefit of using variant here instead of just the base class of which everything should be inheriting from?

Comment (kurtamohler) [reply c10/util/Exception.h:152]: If we just used the base class, then when we copy a warning into a `vector<Warning>`, we lose the information about which subclass it was

File: torch/csrc/Exceptions.cpp [hunk 8]
```diff
@@ -211,21 +211,35 @@ LinAlgError::LinAlgError(const char* format, ...) {
...
 PyWarningHandler::PyWarningHandler() noexcept(true)
-    : prev_handler_(c10::Warning::get_warning_handler()), in_exception_(false) {
-  c10::Warning::set_warning_handler(&internal_handler_);
+    : prev_handler_(c10::WarningUtils::get_warning_handler()),
+      in_exception_(false) {
+  c10::WarningUtils::set_warning_handler(&internal_handler_);
+}
+
+// Get the Python warning type for a warning
+PyObject* map_warning_to_python_type(const c10::Warning& warning) {
```
[the remark this answers is not in the document]
Comment (kurtamohler) [reply torch/csrc/Exceptions.cpp:225]: Great, I was looking for a way to do that and couldn't figure it out. I'll try this

File: c10/util/Exception.h (hunk 7 above)
Comment (albanD) [reply c10/util/Exception.h:152]: But the virtual methods on it will still work fine right? We don't need the type to actually be correct right?

Comment (kurtamohler) [reply c10/util/Exception.h:152]: Unfortunately, no. Here's an example:

<details>
<summary>Click to expand</summary>

```cpp
#include <iostream>
#include <vector>

class Base {
public:
  virtual int id() {
    return 0;
  }
};

class Derived : public Base {
public:
  int id() override {
    return 1;
  }
};

int main() {
  std::vector<Base> buffer;

  Base base;
  Derived derived;

  buffer.push_back(base);
  buffer.push_back(derived);

  std::cout << buffer[0].id() << std::endl;
  std::cout << buffer[1].id() << std::endl;
}
```

</summary>

Running the above code, both `buffer[...].id()` calls return 0, meaning that the base method was called both times

Comment (albanD) [reply c10/util/Exception.h:152]: Ho right, can we make these unique_ptr then? :D

Comment (kurtamohler) [reply c10/util/Exception.h:152]: Yes, smart pointers are an option. There was some discussion earlier in this PR about whether to use smart pointers or variants.

Originally, I did implement this PR with `shared_ptr` (though `unique_ptr` would have been a better choice), and it did work. But one of the the annoying things about it was that to create a `shared_ptr` of the right type you have to call `std::make_shared<subclass_type>` (or `std::make_unique<subclass_type` for `unique_ptr`), where the "subclass_type" here has to actually be the actual subclass type, not the base type. Because of that, in order to have a generic way to create a unique_ptr from any of the warning subclasses, you have to make a template function with explicit overloads for each warning subclass, something like this (code might not be 100% correct):

```cpp
template<>
std::shared_ptr warning_to_shared_ptr(c10::UserWarning warning) {
  return std::make_shared<c10::UserWarning>(warning);
}

// repeat for each warning subclass type
```

Perhaps there was a better way to do it, I'm not sure.

If smart pointers are actually better, then I'm not opposed to switching back. But at the moment it seems to me like using variants makes the implementation simpler.

cc @lezcano

Comment (lezcano) [reply c10/util/Exception.h:152]: My point here is, why do you need to know the type of the objects / why do you need virtual things at all? If they all have a given method, can't you just do a whatever's the equivalent we have to `std::visit` and execute that method for all the elements in the `std::vector`?

Comment (kurtamohler) [reply c10/util/Exception.h:152]: > why do you need to know the type of the objects / why do you need virtual things at all?

I think we have three main options:

1. Keep track of the `c10::Warning` subclass type so that we can map it to the appropriate Python warning type later (whether that's done by the explicit typeid thing currently in this PR or the better way Alban mentioned here: https://github.com/pytorch/pytorch/pull/84101#discussion_r982516484).

2. Add an enum member to `c10::Warning` that we can use to map to the appropriate Python warning type later.

3. Just keep track of the Python warning type directly as a `PyObject*` member of `c10::Warning`, with no need for subclasses.

I went for option 1 because it seemed more consistent with how `c10::Error` works, and it's consistent with the Python side, where warning types are all subclasses of a base class.

Option 2 is mentioned here https://github.com/pytorch/pytorch/pull/84101#issuecomment-1245878625, and I'm not sure if it's better or not, because it seems less consistent with `c10::Error` and Python warnings.

Option 3 is might be the simplest, but I'm not sure how to implement it, and again it seems less consistent.

> If they all have a given method, can't you just do a whatever's the equivalent we have to std::visit and execute that method for all the elements in the std::vector?

I'm having trouble understanding this. What is an example of something that's equivalent to std::visit?

Comment (albanD) [reply c10/util/Exception.h:152]: Consistency with `c10::Error` and python side of things is a good reason for sure.
I am not familiar with the variant approach that's why I was asking. But sounds good.

Comment (lezcano) [reply c10/util/Exception.h:152]: I think either of the 3 options should work. I liked the variant option because it avoids pointers, which is good. Now, I think that @amjames is looking into this, so he'll be helping him finalise this PR :)

**Comment by kurtamohler** [comment]
Andrew described a way to improve the implementation, and I'll give my summary of the idea. I'd like to make sure there is consensus.

The only reason we're doing this refactor is to enable `TORCH_WARN_WITH(warning_type, ...)`, one of the future features in the RFC. Users don't need to instantiate or call methods on `warning_type`. `warning_type` is, from the user's perspective, just a selection between different types of warnings. So, just because errors are polymorphic, that doesn't mean that it's actually "consistent" to try to match it for warnings. The implementations of warnings and errors have to have significant differences anyway, so if there's a simpler and safer way to implement warnings than copying the error class hierarchy, it makes sense to do it that way.

So we can get rid of the polymorphic approach and instead add an enum member to `c10::Warning` that keeps track of which python warning type it is. But it's not actually a normal `enum`, it's an enum-like structure implemented as a `variant`, which I think is what Mario was suggesting in https://github.com/pytorch/pytorch/pull/84101#issuecomment-1246455661 and I didn't understand until now. When we need to pull the warnings out of the buffer and raise them, we use `c10::visit` to get the `PyExc_*Warning`. So we'll only need to call `c10::visit` on the variant in one place.

The variant provides safety because if you add a new warning type to it, the `c10::visit` call site will throw an error if you don't also add its python mapping, and vice versa. Again, Mario mentioned this and I misunderstood.

Here's a simplified example of how this could look. Of course some of the details will look different in reality.

Instead of `print_python_warning_type` from below, we'll have a function that returns the `PyObject*` warning type, which will go in `torch/csrc/Exceptions.cpp`.

```cpp
#include <iostream>
#include <variant>
#include <vector>

namespace c10 {

class Warning {
 public:
  class UserWarning {};
  class DeprecationWarning {};

  using warning_variant_t = std::variant<UserWarning, DeprecationWarning>;

  Warning(warning_variant_t type)
    : type_(type)
  {}

  const warning_variant_t& type() const {
    return type_;
  }

 private:
  warning_variant_t type_;
};

using UserWarning = Warning::UserWarning;
using DeprecationWarning = Warning::DeprecationWarning;

}

template<class... Ts> struct overloaded : Ts... { using Ts::operator()...; };
template<class... Ts> overloaded(Ts...) -> overloaded<Ts...>;

void print_python_warning_type(const c10::Warning& warning) {
  const char* type_name = std::visit(overloaded {
      [](const c10::UserWarning&) { return "PyExc_UserWarning"; },
      [](const c10::DeprecationWarning&) { return "PyExc_DeprecationWarning"; },
  }, warning.type());

  std::cout << type_name << std::endl;
}

int main() {
  auto w0 = c10::Warning(c10::UserWarning());  
  auto w1 = c10::Warning(c10::DeprecationWarning());  

  print_python_warning_type(w0);
  print_python_warning_type(w1);
}
```

cc @amjames @lezcano @albanD

**Comment by albanD** [comment]
Thanks for the details on this!
Just to make sure I follow the implications:
- In this world, the c++ side would be a flat list of warning types while the python side will be a full inheritance hierarchy.
- Won't this cause issue to do the filtering at the c++ level though? A python filter would work on the given class as well as all its subclass. So I feel like we do need the warning hierarchy at the c++ level as well if we want to be able to add the c++ filter in the future.
- `TORCH_WARN_WITH(warning_type, ...)` is actually only an internal construct (at least for the error case it is), the user will only see `TORCH_WARN()` `TORCH_WARN_USER()` TORCH_WARN_DEPRECATION()` etc.
- " The implementations of warnings and errors have to have significant differences anyway" -> why ?

**Comment by amjames** [comment]
> The implementations of warnings and errors have to have significant differences anyway

Maybe *some* differences would be better wording. Warnings don't break control flow, so the have to be stored, possibly with other warnings generated before the appropriate point for them to be reported. These are just fundamental differences between warnings and exceptions, they are similar concepts but in terms of what we must do with them, they are just different. 

The other points you bring up have been considered. The conversation about this was lengthy so I am going to give a sumamry 

**TLDR:** We can't implement C++ warning -> python warning type as a virtual method, so we need to also cary additional info and discriminate on it (like warning has an enum and we switch on it). The variant takes the place of the enum + switch solution that can prevent some types of bugs. It also handles the cases of mapping behaviors to types which might be possible to implement with virtual methods (like filtering). Finally it does not preclude the use of a hierarchy if desired. 

@kurtamohler: some additional ideas came up while writing this you might want to take a look. 

<details>
<summary>Long version</summary>
In both the case of errors and warnings the only information carried by the type is essentially that they are different from one another. Since we handle errors one at a time (we can `catch` the type directly), there is no worry about loss of this information. With warnings a heterogeneous type is required for storage (that is the `warnings_buffer`s owned by several handler implementations). So by the time we handle a warning, if using inheritance the specialized type information is lost, and would need to be communicated by another means such as accessing a `type` member. Using a virtual function for the behavior that needs to discriminate on type is not ideal as at least some of that behavior (generation of appropriate CPython Warning object) is not possible to implement within the class definition. The warning type definition can't rely on python, we can use a stub for a member which will eventually hold the CPython warning object (as mentioned by alban), but at whatever point we initialize that member the warning is already stored as an opaque pointer to base and needs to communicate its specific type by another means.  

So now we have established that at least some behavior of a warning must be implemented external to the class, and have access to discriminating information besides the type itself as it may be abstract at the time handled. A variant is a great solution to this situation.  Some alternatives would be either an enum member, or using `typeid`. The variant is a better option than either alternative as it provides static assurances that any added alternative for the variant is handled by any visitor used, it has some of the pitfalls of a switch + enum solution, but eliminates others. 

The other behavior we have a concrete use case for is filtering. Now filtering could be implemented using virtual method, but I think that would lead to an awkward implementation. What might that look like? 

```c++
void FilteringWarningHandler::process(Warning warning) {
  if (warning.passes_this_filter(filter)){ 
    original_handler->process(warning); 
  } //else {ignore it}
```
Here the warning must inspect the filter, and determine if itself is an allowed type through that filter. Every new filter requires an update to potentially all warning types! 


This would be a more expressive design for that interaction:
```c++
void FilteringWarningHandler::process(Warning warning) {
  if (filter.type_passes(warning.type()){ 
    original_handler->process(warning); 
  } //else {ignore it}
```
Where `filter.type_passes()` is implemented using a visitor. 

There is nothing stopping us from *also* having a hierarchy. The proposed variant solution  would take the place of virtual methods, but having a warning hierarchy is still possible, just not strictly necessary IMO. 

```c++

namespace c10 {
namespace warning_type_flags {
  struct user_warning{};
  struct deprecation_warning{};
  using warning_variant_t = std::variant<user_warning, deprecation_warning>;
}

class Warning {
 public:
  using warning_type_flags::warning_variant_t;

  Warning(warning_variant_t type)
    : type_(type)
  {}

  const warning_variant_t& type() const {
    return type_;
  }
 private:
  warning_variant_t type_;
};

class UserWarning : public Warning {
  public:
    UserWarning() : Warning(warning_type_flags::user_warning()) {};
};

class DeprecationWarning : public Warning {
  public:
    DeprecationWarning() : Warning(warning_type_flags::deprecation_warning) {}
};
```

If custom types were implemented as inheriting from one of these it would also inherit the variant value and therefore still match the same filters. If they need to match both filters on the parent type and filters specific to them only, the tag structs comprising the variant alternatives can also use inheritance. 
</details>

**Comment by kurtamohler** [comment]
Thanks for writing down all your thoughts, @amjames . I think that all makes sense, and I think it answers the questions @albanD brought up.

As Andrew showed, a hierarchical structure doesn't seem necessary, even for when we implement the C++ warning filter in the future. I'll opt for a flat structure for now because it's a little simpler. If we find out that we actually do need a hierarchy later for some reason, it should be fairly straightforward to add it in later.

 Unless there are any other concerns, I'll go ahead and implement this

**Comment by kurtamohler** [comment]
I think this is ready for another review

**Review by albanD — APPROVED** [review]
SGTM!
Thanks for taking care of that!

File: c10/util/Exception.h [hunk 9]
```diff
@@ -112,17 +113,65 @@ class C10_API Error : public std::exception {
...
+      const char* msg,
+      bool verbatim);
+
+  Warning(
+      warning_variant_t type,
+      SourceLocation source_location,
+      ::c10::detail::CompileTimeEmptyString msg,
+      bool verbatim);
+
+  virtual ~Warning();
```
Comment (amjames) [inline c10/util/Exception.h:139]: Warning need not have a virtual destructor any longer.

File: c10/util/Exception.h [hunk 10]
```diff
@@ -112,17 +113,65 @@ class C10_API Error : public std::exception {
   std::string compute_what(bool include_backtrace) const;
 };
 
+class C10_API Warning {
+ public:
+  class C10_API UserWarning {};
+  using warning_variant_t = c10::variant<UserWarning>;
```
Comment (amjames) [inline c10/util/Exception.h:119]: It is probably a good idea to add another warning type, even if we don't exhaustively go through and use it everywhere we should at this point. Having two may uncover some things.

**Review by amjames — COMMENTED** [review]
See inline comments, but elaborating here on one item: 

Adding a second variant alternative an using it in one spot will give us the confidence that the groundwork is laid sufficiently to extend it to any number of warning types after the fact. 

No need to track down and swap out the warning type for all situations that other warning might apply to right now that can be done later. 

Also a question: Do we have any tests of warning machinery that should be updated/augmented?

File: c10/util/Exception.h (hunk 10 above)
Comment (kurtamohler) [reply c10/util/Exception.h:119]: I can add `c10::DeprecationWarning` and use it in one place, with a unit test to make sure the Python translation works.

Just to be clear, this will require adding `TORCH_WARN_DEPRECATION` and `TORCH_WARN_WITH`

Comment (kurtamohler) [reply c10/util/Exception.h]: I added `TORCH_WARN_DEPRECATION`, used it in `torch.ger`, and added a unit test for it

File: test/test_linalg.py [hunk 11]
```diff
@@ -141,6 +142,31 @@ def run_test_case(a, b):
         run_test_case(zero_strided, b)
         run_test_case(a, zero_strided)
 
+    @dtypes(*all_types_and_complex_and(torch.half, torch.bfloat16, torch.bool))
```
Comment (amjames) [inline test/test_linalg.py:145]: I don't think this is a great idea as the deprecated function will disappear at some point.

We would want to fabricate a situation where the warning raised is internal to the test and could be relied upon to not change the test outcome without changes to the warning code. That might be difficult to do.

Comment (kurtamohler) [reply test/test_linalg.py:145]: I'm not 100% sure I understand what you mean, but my impression is that since any normal usage of `TORCH_WARN_DEPRECATED` (and for that matter any `TORCH_WARN*` call) is subject to change, it would be nice to have some more reliable way to check that `TORCH_WARN_DEPRECATED` does the right thing, independent of any other operation.

We could add some function `torch._C._warn_deprecated()`, and others for other warning types, bound in `torch/csrc/autograd/FunctionsManual.cpp`, which just calls `TORCH_WARN*`, and we check that the correct warning type is raised in Python

Comment (albanD) [reply test/test_linalg.py:145]: > I don't think this is a great idea as the deprecated function will disappear at some point.

We actually do this in a couple places and I think that's fine.
As long as you leave a comment there saying explicitly that the particular function is just as an example and it can be swapped if the warning is ever removed.

Comment (kurtamohler) [reply test/test_linalg.py]: I've added `torch._C._warn_deprecation` and `torch._C._warn` with a unit test for them

**Comment by kurtamohler** [comment]
I think this is ready to go unless there are any more thoughts
